### ISSUE TITLE
UPE checkout with free trial subscription

### ISSUE THREAD

**Issue by bborman22** [Automattic/woocommerce-payments#1979]
As noted in this [Github thread](https://github.com/Automattic/woocommerce-payments/issues/1828#issuecomment-850055235), it is not possible to checkout with UPE with a 100% discount. This will be an issue for any purchases with 0 cost. Another example is a subscription with a free trial, where the payment method is needed to be captured.

This issue is currently blocked as we wait for Stripe to implement the ability to SetupIntent using the UPE.

**Comment by bborman22** [comment]
We have [received an update](p1622739701100700/1622738534.100000-slack-C9976E5MJ) that the `SetupIntent` functionality should be available. While we are still waiting on the docs, we should have gotten enough information to get this flow working. #1982 will be the simpler implementation of the `SetupIntent` flow so it may be worth it to wait for that implementation to be finished to draw inspiration from.

The major catch with this issue is that we may mount the UPE on checkout page load with a regular payment intent. Then through some action on the checkout page, the checkout may become for an amount of 0. When this happens we will need to use a setup intent instead, which may require re-mounting the UPE with the newly created setup intent. In short, this issue will need a bit of research and experimentation to find the right flow to allow checkout with UPE for an amount of 0.

**Comment by FangedParakeet** [comment]
@bborman22 I've been testing this one out yesterday (while working on #2057 to see what happens when there is no charge) and today with a 100% discount coupon. When you apply the coupon, the entire payment gateway section gets removed from the checkout and after you checkout there is no payment method listed with the completed order. The WC Checkout class doesn't even call the payment gateway either, unless someone meddles with either the `woocommerce_order_needs_payment` or `woocommerce_cart_needs_payment` filters. 

However, in either case, looking at the previous WCPay gateway, it looks like the only scenario handled when payment is not needed is the case of a subscription payment with no initial payment. Is there a separate issue out there to cover that case or do you want to use this issue to tackle using `SetupIntent`s to handle recurring subscription payments without an initial fee?

**Comment by bborman22** [comment]
Thanks for taking a look at that, that description matches what I saw the other day as well. If it wouldn't be too confusing, I think repurposing this issue to apply to the subscription scenario works for me? That would keep some of this conversation all in one place. I don't feel strongly either way though.

**Comment by FangedParakeet** [comment]
Cool, we can keep the convo here, no complaints. Is this blocked by #1927 at all or can this train get rolling on its own?

**Comment by bborman22** [comment]
True, it will be blocked by #1927 since we can't really test out the subscription flow until that's done. I'm making more progress on it today, hopeful to be ready sometime Monday.

**Comment by bborman22** [comment]
With the merge of the Subcription changes in #1927, this item is now unblocked.
